### PR TITLE
Dataset download entrypoint

### PR DESCRIPTION
**Pull request by jankrepl**

### CODE REVIEW

File: src/atldld/sync.py [hunk 1]
```diff
@@ -260,59 +227,190 @@ def download_parallel_dataset(
...
+        self.include_expression = include_expression
+        self.downsample_img = downsample_img
+
+        self.metadata: Dict[str, Any] = {}
+        # populated by calling `fetch_metadata`
+
+    def __len__(self) -> int:
+        """Return the number of images in the dataset."""
+        if "images" not in self.metadata:
+            self.fetch_metadata()
```
Comment (jankrepl) [inline src/atldld/sync.py:252]: Heh, I would raise an error instead and only fetch the metadata when the user explicitly calls `fetch_metadata()`

Anyway, it is a detail and a matter of taste I guess. Your call.

File: src/atldld/sync.py [hunk 2]
```diff
@@ -260,59 +227,190 @@ def download_parallel_dataset(
...
+            Generator yielding consecutive four tuples of
+            (image_id, constant_ref_coordinate, img, df).
+            The `constant_ref_coordinate` is the dimension in the given axis in microns.
+            The `img` is the raw gene expression image with dtype `uint8`.
+            The `df` is the displacement field.
+            Note that the sorting. If `include_expression=True` then last returned image
+            is the processed expression image.
+            That is the generator yield (image_id, p, img, df, img_expr).
+        """
+        if self.metadata is None:
```
Comment (jankrepl) [inline src/atldld/sync.py:358]: This does not seem to be consistent with the current constructor which initialized `self.metadata = {}`

Comment (jankrepl) [reply src/atldld/sync.py:358]: Also it seems like there is no unit test in case the metadata is empty

File: tests/conftest.py [hunk 3]
```diff
@@ -29,11 +29,18 @@
...
+    """Return data folder."""
+    return DATA_FOLDER
+
+
 @pytest.fixture(autouse=True)
 def custom_cache_dir(monkeypatch, tmpdir):
     # Automatically use a custom cache directory for all tests to avoid writing
     # cache data to the user's real cache.
     monkeypatch.setenv("XDG_CACHE_HOME", str(tmpdir))
+    return pathlib.Path(tmpdir) / "atldld"
```
Comment (jankrepl) [inline tests/conftest.py:43]: I think this is not necessary.

If you want to actually have this folder inside of tests you can just call `atldld.config.user_cache_dir` and you should get it. This works since it is reading from the `XDG_CACHE_HOME`.

File: tests/test_sync.py [hunk 4]
```diff
@@ -92,56 +87,54 @@ def test_local_equals_API(self, pir_to_xy_response, downsample_ref):
...
+        # Mocking the responses to request
+        with open(data_folder / "sync" / "SectionDataSet.txt") as f:
+            body_dataset = f.read()
+
+        responses.add(
+            responses.GET,
+            "https://api.brain-map.org/api/v2/data/query.json?"
+            "criteria=model::SectionDataSet,rma::criteria,[id$eq123],"
+            "rma::include,alignment3d,rma::options[start_row$eq0][num_rows$eq25000]",
+            body=body_dataset,
```
Comment (jankrepl) [inline tests/test_sync.py:110]: I guess another option here would be to patch the `rma_all` function and not worry about what is happening inside of it.

The current approach kind of assumes that we see both inside of `DatasetDownloader` and `rma_all` code. Which means that if we make changes either inside of `rma_all` or `DatasetDownloder` then this test will fail.

I would lean towards just patching `rma_all` and thus this test would only fail if we introduce changes inside of `DatasetDownloader`. Also, if we only patch `rma_all` then we just need to return the minimal metadata necessary to make the `DatasetDownloader` work. Anyway, I can totally help with this since it would require some work + refactoring.

**Review by jankrepl — COMMENTED** [review]
A lot of great work!

I left some comments in the code. Overall, I think it look great and I would only write a couple of additional unit tests to make sure the `sync.py` is 100% covered.

File: tests/test_sync.py [hunk 5]
```diff
@@ -92,56 +87,54 @@ def test_local_equals_API(self, pir_to_xy_response, downsample_ref):
...
-        # Patching
-        monkeypatch.setattr("atldld.sync.get_2d_bulk", get_2d_bulk_fake)
-        monkeypatch.setattr("atldld.sync.get_3d", get_3d_fake)
-        monkeypatch.setattr("atldld.sync.get_image", get_image_fake)
-        monkeypatch.setattr("atldld.sync.CommonQueries", common_queries_fake)
-        monkeypatch.setattr("atldld.sync.xy_to_pir_API_single", xy_to_pir_fake)
+        # Save fake image in cache folder
+        custom_cache_dir.mkdir(exist_ok=True)
+        fake_img = Image.fromarray(np.zeros((100, 200), dtype=np.uint8))
+        fake_img.save(custom_cache_dir / "101945191-0.jpg")
```
Comment (jankrepl) [inline tests/test_sync.py:128]: Similar to my comment above. I think it is better (+ simpler) if we patch `atldld.utils.get_image` directly. This is more in the spirit of unit testing, since in this test we are focusing on `DatasetDownloader` and we can make an assumption that "atldld.utils.get_image" works as expected.

File: src/atldld/cli.py [hunk 6]
```diff
@@ -167,6 +168,105 @@ def dataset_info(dataset_id):
...
+    Plane of section         : {downloader.metadata["dataset"]["plane_of_section_id"]}
+    """
+    click.secho(textwrap.dedent(additional_info).strip(), fg="green")
+
+    metadata = {
+        "dataset_id": dataset_id,
+        "downsample_ref": downsample_ref,
+        "downsample_img": downsample_img,
+        "per_image": {},
+    }
```
[the remark this answers is not in the document]
Comment (EmilieDel) [reply src/atldld/cli.py:248]: It is a good point, maybe we can open a ticket to analyse which data are useful and add them if it is not the case yet.

File: src/atldld/cli.py [hunk 7]
```diff
@@ -167,6 +168,105 @@ def dataset_info(dataset_id):
...
+            plt.imsave(str(img_path), img_synced)
+
+            if img_expr is not None:
+                img_expr_synced = df.warp(img_expr)
+                img_expr_path = output_folder / f"{image_id}_expr.png"
+                plt.imsave(str(img_expr_path), img_expr_synced)
+
+            metadata["per_image"][image_id] = {
+                "section_coordinate": section_coordinate,
+            }
```
[the remark this answers is not in the document]
Comment (EmilieDel) [reply src/atldld/cli.py:264]: Same comment as above.

File: src/atldld/sync.py [hunk 8]
```diff
@@ -260,59 +227,190 @@ def download_parallel_dataset(
...
+        include_expression: bool = False,
+        downsample_img: int = 0,
+    ):
+        self.dataset_id = dataset_id
+        self.downsample_ref = downsample_ref
+        self.detection_xy = detection_xy
+        self.include_expression = include_expression
+        self.downsample_img = downsample_img
+
+        self.metadata: Dict[str, Any] = {}
```
[the remark this answers is not in the document]
Comment (EmilieDel) [reply src/atldld/sync.py:246]: Doing so raises issue elsewhere:
```
lint run-test: commands[4] | mypy setup.py src/atldld tests docs
src/atldld/cli.py:238: error: Value of type "Optional[Dict[str, Any]]" is not indexable  [index]
src/atldld/cli.py:239: error: Value of type "Optional[Dict[str, Any]]" is not indexable  [index]
```
I know that `self.metadata = None` really brings the info that we never fetch the metadata, but I think it is easier to instantiate it as an empty `dict` in terms of typing. What do you think ?

[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py:246]: `mypy` does not seem to realize that it is going to be turned into `dict` inside of `fetch_matadata`.
Honestly, `mypy` is a blackbox to me:D Anyway, let's keep it the way it is (`self.metadata = {}`).

File: src/atldld/cli.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/cli.py:248]: Sounds good:)

File: src/atldld/sync.py [hunk 9]
```diff
@@ -260,59 +227,190 @@ def download_parallel_dataset(
...
+        1. Get metadata for the entire dataset (e.g. `affine_3d`)
+        2. Get metadata for all images inside of the dataset (e.g. `affine_2d`)
+
+        Parameters
+        ----------
+        force_redownload
+            If yes, force to redownload the metadata. Otherwise, if
+            the metadata have been computed once, they are not computed again.
+        """
+        if self.metadata != {} and not force_redownload:
```
[the remark this answers is not in the document]
Comment (EmilieDel) [reply src/atldld/sync.py]: Done in feb3b63.

File: src/atldld/sync.py (hunk 1 above)
Comment (EmilieDel) [reply src/atldld/sync.py]: Done in feb3b63.

File: src/atldld/sync.py (hunk 2 above)
Comment (EmilieDel) [reply src/atldld/sync.py]: Done in feb3b63 (first comment).

File: tests/conftest.py (hunk 3 above)
Comment (EmilieDel) [reply tests/conftest.py]: Good point. Done in cae39ef

File: tests/test_sync.py (hunk 4 above)
Comment (EmilieDel) [reply tests/test_sync.py]: Patched `rma_all` in cae39ef.

File: tests/test_sync.py (hunk 5 above)
Comment (EmilieDel) [reply tests/test_sync.py]: Patched `get_image` in cae39ef.

File: tests/conftest.py (hunk 3 above)
Comment (jankrepl) [reply tests/conftest.py]: Perfect!

File: tests/test_sync.py [hunk 10]
```diff
@@ -15,13 +15,15 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 """Test for sync.py module."""
-from unittest.mock import Mock
+import json
+from unittest.mock import MagicMock
```
Comment (jankrepl) [inline tests/test_sync.py:19]: I think it is enough to only use `Mock`

```suggestion
from unittest.mock import Mock
```

File: src/atldld/sync.py (hunk 2 above)
Comment (EmilieDel) [reply src/atldld/sync.py]: Done in 632f947 (second comment).

File: tests/test_sync.py [hunk 11]
```diff
@@ -92,105 +94,122 @@ def test_local_equals_API(self, pir_to_xy_response, downsample_ref):
...
+                    data = json.load(f)
+                return [
+                    data,
+                ]  # In a list because can have several images
+            else:
+                raise ValueError("Not expected")
+
+        mock = MagicMock(side_effect=mock_rma_all)
+        monkeypatch.setattr("atldld.sync.rma_all", mock)
+        monkeypatch.setattr("atldld.utils.get_image", np.zeros(grid_shape))
```
Comment (jankrepl) [inline tests/test_sync.py:141]: `monkeypatch.setattr("atldld.utils.get_image", np.zeros(grid_shape))` does not work (at least on my computer). I think one always needs to specify the module that imports the function (`sync`) rather than the one where it is implemented (`utils`).

```suggestion
        fake_rma_all = Mock(side_effect=mock_rma_all)
        fake_get_image = Mock(return_value=np.zeros(grid_shape))
        monkeypatch.setattr("atldld.sync.rma_all", fake_rma_all)
        monkeypatch.setattr("atldld.sync.get_image", fake_get_image)
```

I also change the variable naming since there are 2 mocks

Comment (EmilieDel) [reply tests/test_sync.py]: Done in 48d4b15. Thanks a lot for catching this!

File: tests/test_sync.py (hunk 10 above)
Comment (EmilieDel) [reply tests/test_sync.py]: Done in 48d4b15.

File: src/atldld/sync.py [hunk 12]
```diff
@@ -260,59 +227,190 @@ def download_parallel_dataset(
...
+                    "section_number": r_image["section_number"],
+                }
+            )
 
-        img = get_image(image_id, downsample=downsample_img)
-
-        if not include_expression:
-            yield image_id, slice_ref_coordinate, img, df
+        images.sort(key=lambda x: -x["section_number"])  # type: ignore
+        metadata["images"] = images  # type: ignore[assignment]
```
Comment (Stannislav) [inline src/atldld/sync.py:322]: I think one can avoid both type ignores like this (I haven't tried though):
1. In the lambda cast explicitly to `int`:
```python
images.sort(key=lambda x: -int(x["section_number"]))
```
2. Instead of defining the `metadata` dict and than later assigning `metadata["images"]` define `metadata` later, when `images` already exists:
  ```python
  metadata = {
      "id": ...,
      "images": ...,
      ...
  }
  ```

Comment (EmilieDel) [reply src/atldld/sync.py]: It seems it is working fine (11995ac). Thanks for the suggestions!

**Comment by EmilieDel** [comment]
@jankrepl @Stannislav Should we merge this PR ?

**Comment by EmilieDel** [comment]
@jankrepl @Stannislav Should we merge this PR ?
